plaid/src/executor/mod.rs

```diff
@@ -316,8 +317,50 @@ fn execution_loop(
             Some(module) => module,
         };
 
+        // If ANY of the modules that operate on this log type cannot be executed in parallel, we need to check
+        // to see if any of them are currently being executed
+        if execution_modules.iter().any(|module| module.is_locked()) {
+            // If this rule is currently being processed, send the message back into the execution queue
+            // This allows us to guarantee that the rule will only be executed on 1 thread at any given time
+
+            info!(
+                "Message of type [{}] is currently being processed by another rule. Requeueing the message for later execution.",
+                message.type_
+            );
+
+            if let Err(e) = sender.send(message.clone()) {
+                error!(
+                    "Failed to requeue message of type [{}]. It is permanently lost. Error: {e}",
+                    message.type_
+                );
+            }
+
+            continue;
+        };
+
         // For every module that operates on that log type
         for plaid_module in execution_modules {
+            // Mark this rule as currently being processed by locking the mutex
+            // This lock will be dropped at the end of the iteration so we don't
+            // need to handle unlocking it
+            let _lock = match plaid_module.concurrency_unsafe {
+                Some(ref mutex) => match mutex.lock() {
```

Comment:
I believe this can deadlock if Plaid ever decided to execute rules in random orders (maybe for performance reasons, or because in the future the rules structure can be updated at runtime).

That could allow some mutexes to be acquired, while waiting for others. Happening between two threads, they could lock and prevent log processing.